Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.h

```diff
@@ -236,5 +308,181 @@ namespace Terrain
+                }
+            };
+
+            // Create the job and start it immediately.
+            AZ::Job* processJob = AZ::CreateJobFunction(jobFunction, true, jobContext);
+            processJob->Start();
+        }
+    }
+
+    template<typename PerSurfacePointFunctionType>
+    inline void TerrainSystem::ProcessFromRegionAsync(PerSurfacePointFunctionType perSurfacePointFunction,
+        const AZ::Aabb& inRegion,
+        const AZ::Vector2& stepSize,
+        AzFramework::Terrain::SurfacePointRegionFillCallback perPositionCallback,
+        Sampler sampleFilter,
+        AZStd::shared_ptr<ProcessAsyncParams> params) const
+    {
+        if (!perPositionCallback)
+        {
+            return;
+        }
+
+        // ToDo: Determine the number of jobs to split the work into based on:
+        // 1. The number of available worker threads.
+        // 2. The desired number of jobs as passed in.
+        // 3. The size of the area being processed.
+        //
+        // Note: We are currently restricting the number of worker threads to one
+        // because splitting the work over multiple threads causes contention when
+        // locking various mutexes, resulting in slower overall wall time for async
+        // requests split over multiple threads vs one where all the work is done on
+        // a single thread. The latter is still preferable over a regular synchronous
+        // call because it is just as quick and prevents the main thread from blocking.
+        // Once the mutex contention issues have been addressed, we should come up with
+        // an algorithm to break up 'inRegion' into sub-regions (or lists of positions?)
+        // so that async calls automatically split the work between available job manager
+        // worker threads, unless the ProcessAsyncParams specifiy a desired number of jobs.
+        const int32_t numWorkerThreads = 1; // m_terrainJobManager->GetNumWorkerThreads();
+        const int32_t numJobsDesired = params ? params->m_desiredNumberOfJobs : -1;
+        const int32_t numJobs = (numJobsDesired > 0) ? AZStd::min(numWorkerThreads, numJobsDesired) : numWorkerThreads;
+        if (numJobs <= 0)
+        {
+            AZ_Warning("TerrainSystem", false, "No positions to process.");
+            return;
+        }
+
+        // Determine the desired terrain job context so the jobs can be cancelled.
+        TerrainJobContext* jobContext = (params && params->m_terrainJobContext) ?
+                                        params->m_terrainJobContext.get() :
+                                        m_defaultTerrainJobContext.get();
+
+        // Split the work across multiple jobs.
+        AZ_Assert(numJobs == 1, "We don't yet support breaking up regions."); // Temp until we figure out how to break up the region.
+        AZStd::shared_ptr<AZStd::atomic_int> numJobCompletionsRemaining = AZStd::make_shared<AZStd::atomic_int>(numJobs);
```

Comment:
At the moment it cannot, because a single TerrainJobContext can be used to make multiple Async calls. With the change to make each request return it's own unique TerrainJobContext it could be moved there, the only downside I see to that is it opens the door for the calling code to mess with it?

Reply:
It could get moved to protected / private with TerrainSystem as a friend class to prevent anyone else from messing with it, maybe. Maybe it's not worth it, but the value seemed related to the job context itself, which is why I was thinking it just logically belonged there. Nothing I feel strongly about though.

Reply:
The friend class isn't really a good option because this is in a different module and namespace to the TerrainSystem that needs to poke it, but I did it anyway and just added an 'OnJobCompleted' function that decrements the count and returns true if it happens to the last job associated with the context to be completed.